flink-yarn/src/test/java/org/apache/flink/yarn/YarnApplicationFileUploaderTest.java

```diff
@@ -73,6 +74,38 @@ void testRegisterProvidedLocalResources(@TempDir File flinkLibDir) throws IOExce
         }
     }
 
+    @Test
+    void testRegisterProvidedLocalResourcesWithParentDir(@TempDir File flinkLibDir)
+            throws IOException {
+        final Map<String, String> filesWithParentDir = getFilesWithParentDir();
+
+        generateFilesInDirectory(flinkLibDir, filesWithParentDir);
+
+        try (final YarnApplicationFileUploader yarnApplicationFileUploader =
+                YarnApplicationFileUploader.from(
+                        FileSystem.get(new YarnConfiguration()),
+                        new Path(flinkLibDir.toURI()),
+                        Collections.singletonList(new Path(flinkLibDir.toURI())),
+                        ApplicationId.newInstance(0, 0),
+                        DFSConfigKeys.DFS_REPLICATION_DEFAULT)) {
+
+            List<String> classPath = yarnApplicationFileUploader.registerProvidedLocalResources();
+
+            Set<String> expectedClassPathEntries = new HashSet<>();
+            Set<String> resources = new HashSet<>();
+            Set<String> resourcesDir = new HashSet<>();
+            for (String filePath : filesWithParentDir.keySet()) {
+                String parentDir = new Path(filePath).getParent().toString();
+                resourcesDir.add(parentDir);
+                resources.add(filePath);
+            }
+            resourcesDir.stream().sorted().forEach(expectedClassPathEntries::add);
+            resources.stream().sorted().forEach(expectedClassPathEntries::add);
+
+            assertThat(classPath).containsExactlyInAnyOrderElementsOf(expectedClassPathEntries);
```

Comment:
how are we ensuring that the classpath is in expected order? `Set` doesn't guarantee ordering therefore `expectedClassPathEntries` should be `List<String>` instead of `Set<String>`. Also I prefer to manually construct the `expectedClassPathEntries` rather than programmatically constructing it for stronger validation.

Reply:
Missed using List instead of Set. Updated it and also manually constructing the classpath.